core/src/main/java/org/apache/iceberg/Partitioning.java

```diff
@@ -195,41 +198,68 @@ public Void alwaysNull(int fieldId, String sourceName, int sourceId) {
   }
 
   /**
-   * Builds a common partition type for all specs in a table.
+   * Builds an intersection of all partition types in a table.
    *
-   * <p>Whenever a table has multiple specs, the partition type is a struct containing all columns
-   * that have ever been a part of any spec in the table.
+   * <p>Whenever a table has multiple specs, the common partition type is a struct containing only
+   * fields that are present in every spec of the table. In other words, the struct fields represent
+   * an intersection of all partition types.
```

Comment:
Empty struct if we have no common partitionType?

Reply:
I think it may help to also note that if the table contains a single spec it will just be returned.

Reply:
I throw an exception if there is no common partition type for now. Let me think on returning an empty struct.

Reply:
I think either are valid ... but I think we tend in utility methods to rely on the caller to decided whether or not output of the method is valid

Reply:
I feel we either need one more method to check if a common type exists or return an empty struct from this one.

Reply:
Looking below the behavior is 

Common(2 specs with something in common) Struct of common fields
Common( two specs with nothing in common ) Exception
Common( 1 spec) " Return passed in Spec
Common(0 specs) return empty spec

I think we probably should just return the empty spec for the validation failure mode

Reply:
Updated.